src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -132,8 +145,14 @@ private void EmitParameter(ParameterSymbol parameterSymbol)
             writer.WriteEndObject();
         }
 
-        private void EmitVariables()
+        private void EmitVariablesIfPresent()
         {
+            if (this.context.SemanticModel.Root.VariableDeclarations.Length == 0)

```

Comment:
The reason why it still sometimes emits empty variables in the JSON is because some bicep variables require to be inlined if their value is an expression that has a `listKeys()` or `reference()` function (those functions can't be used in variables in JSON).

Can you change the condition to the following (writing from memory and didn't validate, so YMMV):
```csharp
if (!this.context.SemanticModel.Root.VariableDeclarations.Any(symbol => !this.context.VariablesToInline.Contains(symbol)))
{
  ...
}
```

In addition, can you update the `EmitterContext` class to change the `VariablesToInline` property to an `ImmutableHashSet<T>`?

Reply:
Done and your memory served you pretty well as it did the trick. Thanks for the explanation too. Good to know and get my head around how this works.